### PR TITLE
Remove @0xproject/deployer.Deployer. Make contracts able to deploy themselves

### PR DESCRIPTION
**Pull request by LogvinovLeon**
## Description
 Remove `@0xproject/deployer.Deployer` Rename the package to `@0xproject/sol-compiler`. Make the contracts able to deploy themselves.

## Motivation and Context

This allows us to make constructor accept ABI v2 structs and be fully type-safe.
Also - there is now a more clear separation of concerns.

Instead of:
```typescript
import { Deployer, Compiler } from '@0xproject/deployer';
```
It's now:
```typescript
import { Compiler } from '@0xproject/sol-compiler';
```
And:
```typescript
await DummyToken.deployAsync(abi, bytecode, provider);
```

## How Has This Been Tested?

All tests pass

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [x] New feature (non-breaking change which adds functionality)
*   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [x] Change requires a change to the documentation.
*   [x] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [x] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [x] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/0x.js/package.json [hunk 1]
```diff
@@ -16,15 +16,14 @@
         "prebuild": "run-s clean generate_contract_wrappers",
         "build": "run-p build:umd:prod build:commonjs; exit 0;",
         "generate_contract_wrappers": "node ../abi-gen/lib/index.js --abis 'src/compact_artifacts/@(Exchange|Token|TokenTransferProxy|EtherToken|TokenRegistry|DummyToken).json' --template ../contract_templates/contract.handlebars --partials '../contract_templates/partials/**/*.handlebars' --output src/contract_wrappers/generated --backend ethers && prettier --write 'src/contract_wrappers/generated/**.ts'",
-        "lint": "tslint --project . 'src/**/*.ts' 'test/**/*.ts'",
+        "lint": "tslint --project .",
```
Comment (LogvinovLeon) [inline packages/0x.js/package.json]: It's redundant. It takes the globs from tslint.json

File: packages/0x.js/package.json [hunk 2]
```diff
@@ -16,15 +16,14 @@
...
         "test:circleci": "run-s test:coverage",
         "test": "run-s clean test:commonjs",
         "test:coverage": "nyc npm run test --all && yarn coverage:report:lcov",
         "coverage:report:lcov": "nyc report --reporter=text-lcov > coverage/lcov.info",
-        "update_artifacts": "for i in ${npm_package_config_contracts}; do copyfiles -u 4 ../migrations/artifacts/1.0.0/$i.json test/artifacts; done;",
         "clean": "shx rm -rf _bundles lib test_temp scripts",
         "build:umd:prod": "NODE_ENV=production webpack",
-        "build:commonjs": "tsc && yarn update_artifacts && copyfiles -u 2 './src/compact_artifacts/**/*.json' ./lib/src/compact_artifacts && copyfiles -u 3 './lib/src/monorepo_scripts/**/*' ./scripts",
+        "build:commonjs": "tsc && copyfiles -u 2 './src/compact_artifacts/**/*.json' ./lib/src/compact_artifacts && copyfiles -u 3 './lib/src/monorepo_scripts/**/*' ./scripts",
```
Comment (LogvinovLeon) [inline packages/0x.js/package.json]: We now don't copy artifacts but point to ones in migrations

File: packages/0x.js/package.json [hunk 3]
```diff
@@ -34,7 +33,6 @@
     },
     "config": {
         "compact_artifacts": "Exchange DummyToken ZRXToken Token EtherToken TokenTransferProxy TokenRegistry",
-        "contracts": "Exchange DummyToken ZRXToken Token WETH9 TokenTransferProxy MultiSigWallet MultiSigWalletWithTimeLock MultiSigWalletWithTimeLockExceptRemoveAuthorizedAddress MaliciousToken TokenRegistry Arbitrage EtherDelta AccountLevels",
```
Comment (LogvinovLeon) [inline packages/0x.js/package.json]: Was unused

File: packages/0x.js/test/global_hooks.ts [hunk 4]
```diff
@@ -1,7 +1,16 @@
...
+import * as path from 'path';
 
-import { deployer } from './utils/deployer';
+import { constants } from './utils/constants';
+import { provider } from './utils/web3_wrapper';
 
-before('migrate contracts', async () => {
-    await runMigrationsAsync(deployer);
+before('migrate contracts', async function() {
+    this.timeout(20000);
```
Comment (LogvinovLeon) [inline packages/0x.js/test/global_hooks.ts]: Migration takes time

File: packages/base-contract/src/index.ts [hunk 5]
```diff
@@ -37,9 +40,40 @@ export class BaseContract {
...
+                stateMutability: 'nonpayable',
+                payable: false,
+                inputs: [],
+            };
+        }
+    }
     protected static _bnToBigNumber(type: string, value: any): any {
         return _.isObject(value) && value._bn ? new BigNumber(value.toString()) : value;
     }
+    protected static async _applyDefaultsToDeployTxDataAsync<T extends Partial<TxData | TxDataPayable>>(
```
Comment (LogvinovLeon) [inline packages/base-contract/src/index.ts]: We needed a static version for a constructor

File: packages/base-contract/src/index.ts [hunk 6]
```diff
@@ -37,9 +40,40 @@ export class BaseContract {
...
+        txData: T,
+        defaults: Partial<TxData>,
+        estimateGasAsync?: (txData: T) => Promise<number>,
+    ): Promise<TxData> {
+        const txDataWithDefaults: TxData = {
+            ...defaults,
+            ...(txData as any),
```
Comment (LogvinovLeon) [inline packages/base-contract/src/index.ts]: Typescript can't proove that it's spreadable. Look at the comment in the function bellow

File: packages/contract_templates/partials/tx.handlebars [hunk 7]
```diff
@@ -33,7 +33,7 @@ public {{this.tsName}} = {
     ): Promise<number> {
         const self = this as any as {{contractName}}Contract;
         const inputAbi = self._lookupAbi('{{this.functionSignature}}').inputs;
-        [{{> params inputs=inputs}}] = BaseContract._formatABIDataItemList(inputAbi, [{{> params inputs=inputs}}], BaseContract._bigNumberToString.bind(this));
+        [{{> params inputs=inputs}}] = BaseContract._formatABIDataItemList(inputAbi, [{{> params inputs=inputs}}], BaseContract._bigNumberToString);
```
Comment (LogvinovLeon) [inline packages/contract_templates/partials/tx.handlebars]: No need to bind static function

File: packages/migrations/src/migration.ts [hunk 8]
```diff
@@ -1,36 +1,67 @@
...
  * the migration should be written to run synchronously.
- * @param deployer Deployer instance.
+ * @param provider  Web3 provider instance.
+ * @param artifactsDir The directory with compiler artifact files.
+ * @param defaults Default transaction values to use when deploying contracts.
  */
-export const runMigrationsAsync = async (deployer: Deployer) => {
-    const web3Wrapper: Web3Wrapper = deployer.web3Wrapper;
-    const accounts: string[] = await web3Wrapper.getAvailableAddressesAsync();
+export const runMigrationsAsync = async (provider: Provider, artifactsDir: string, defaults: Partial<TxData>) => {
```
Comment (fabioberger) [inline packages/migrations/src/migration.ts]: Let's rename `defaults` to `txDefaults`

File: packages/sol-cov/package.json [hunk 9]
```diff
@@ -54,10 +54,9 @@
         "solidity-parser-antlr": "^0.2.8"
     },
     "devDependencies": {
-        "@0xproject/deployer": "^0.3.5",
```
Comment (fabioberger) [inline packages/sol-cov/package.json]: Add `sol-compiler` here

File: packages/website/ts/containers/sol_compiler_documentation.ts [hunk 10]
```diff
@@ -12,31 +12,29 @@ import { constants } from 'ts/utils/constants';
...
-    deployer: 'deployer',
     types: docConstants.TYPES_SECTION_NAME,
 };
 
 const docsInfoConfig: DocsInfoConfig = {
-    id: DocPackages.Deployer,
+    id: DocPackages.SolCompiler,
     type: SupportedDocJson.TypeDoc,
-    displayName: 'Deployer',
+    displayName: 'Sol Compiler',
```
Comment (fabioberger) [inline packages/website/ts/containers/sol_compiler_documentation.ts]: `Solidity Compiler`

File: packages/website/ts/containers/sol_compiler_documentation.ts [hunk 11]
```diff
@@ -45,13 +43,12 @@ const docsInfoConfig: DocsInfoConfig = {
...
-        [docSections.compiler]: ['"deployer/src/compiler"'],
-        [docSections.deployer]: ['"deployer/src/deployer"'],
-        [docSections.types]: ['"deployer/src/utils/types"', '"types/src/index"'],
+        [docSections.compiler]: ['"sol-compiler/src/compiler"'],
+        [docSections.types]: ['"sol-compiler/src/utils/types"', '"types/src/index"'],
     },
     menuSubsectionToVersionWhenIntroduced: {},
     sections: docSections,
-    visibleConstructors: [docSections.compiler, docSections.deployer],
+    visibleConstructors: [docSections.compiler],
```
Comment (fabioberger) [inline packages/website/ts/containers/sol_compiler_documentation.ts]: Below we have `Web3: constants.URL_WEB3_DOCS`. Is Web3 still part of the public interface of `sol-compiler`?

Comment (fabioberger) [reply packages/website/ts/containers/sol_compiler_documentation.ts]: What about `ContractInstance`?

File: packages/website/ts/index.tsx [hunk 12]
```diff
@@ -86,7 +86,10 @@ render(
                             <Route path={WebsitePaths.Wiki} component={Wiki as any} />
                             <Route path={`${WebsitePaths.ZeroExJs}/:version?`} component={LazyZeroExJSDocumentation} />
                             <Route path={`${WebsitePaths.Connect}/:version?`} component={LazyConnectDocumentation} />
-                            <Route path={`${WebsitePaths.Deployer}/:version?`} component={LazyDeployerDocumentation} />
+                            <Route
+                                path={`${WebsitePaths.SolCompiler}/:version?`}
+                                component={LazySolCompilerDocumentation}
+                            />
```
Comment (fabioberger) [inline packages/website/ts/index.tsx]: We should keep the `/docs/deployer` url around and re-direct it to `docs/sol-compiler`. See `Legacy endpoints` section below.

File: packages/sol-cov/package.json (hunk 9 above)
Comment (LogvinovLeon) [reply packages/sol-cov/package.json]: Addressed offline

**Pull request merged**